File: RACE/RACE_report.py
import pandas as pd
import re
from pathlib import Path
from janitor import clean_names


# Path
# path = Path('datasets/home-dataset/data/')
path = Path(__file__).parent


# Functions
def read_excel_file(filename):
    df = pd.read_excel(
        filename,
        sheet_name="Query",
        skiprows=11,
        # dtype={"ConsUnit": str, "Plant": str, "Outlet": str},
    )
    # change data type
    df = df.astype({"ConsUnit": str, "Plant": str, "Outlet": str})

    # change column names
    df = df.rename(
        columns={
            "Unnamed: 1": "FS item description",
            "Unnamed: 3": "CU name",
            "Unnamed: 5": "Plant name",
            "Unnamed: 7": "Outlet name",
            "YTD - 1": "YTD PM",  # PM means Previous Month
        }
    )
    df = df.rename(columns=lambda x: re.sub("\nACT", "", x))
    # clean column names
    df = clean_names(df)
    return df


def process_currency(df, lc_gc):
    df = df.assign(currency=lc_gc)
    df = df[["currency"] + [col for col in df.columns if col != "currency"]]
    return df


def outlet():
    # POC
    col_poc = ["division", "bu", "new_outlet", "new_outlet_name"]

    df = pd.read_excel(path / "meta" / "New outlet.xlsx", usecols="A:F", dtype="str")
    df = clean_names(df)
    df = df.drop(columns=["outlet_name"])
    df = df[["outlet"] + col_poc]
    return df


def join_with_outlet(df):
    outlet_df = outlet()
    df = df.merge(outlet_df, on="outlet", how="left")
    return df


# Split P&L and Balance sheet
def profit_and_loss(df):
    cols_to_drop = ["period_0"] + [f"ytd_{i}" for i in range(13)]
    df = df.drop(cols_to_drop, axis=1)
    df = df.loc[df["financial_statement_item"].str.contains("^3|^CO")]
    return df


def balance_sheet(df):
    cols_to_drop = [f"period_{i}" for i in range(13)]
    df = df.drop(cols_to_drop, axis=1)
    df = df.loc[df["financial_statement_item"].str.contains("^1|^2")]
    return df


# Input data
path_lc = path / "data" / "Analysis FS Item Hierarchy for CU 698_LC.xlsx"
path_gc = path / "data" / "Analysis FS Item Hierarchy for CU 698_GC.xlsx"


# Combine data
lc = read_excel_file(path_lc)
lc = process_currency(lc, "LC")

gc = read_excel_file(path_gc)
gc = process_currency(gc, "GC")

race = pd.concat([lc, gc])
race = join_with_outlet(race)


# Split data
race_pnl = profit_and_loss(race)
race_bs = balance_sheet(race)


# Output data
race_pnl.to_csv(path / "output" / "RACE Profit and Loss.csv", index=False, na_rep="0")
race_bs.to_csv(path / "output" / "RACE Balance sheet.csv", index=False, na_rep="0")

File: PnL/src/dplyr/RACE report.R
library(dplyr)
library(readr)
library(stringr)
library(readxl)
library(janitor)
library(here)


# Path
path <- here("PnL")


# Functions
read_excel_file <- function(filename, version) {
  df <- read_excel(filename, sheet = "Query", skip = 11) |>
    rename(c(
      `FS item description` = "...2",
      `CU name` = "...4",
      `Plant name` = "...6",
      `Outlet name` = "...8",
      `YTD PM` = "YTD - 1"
    )) |>
    rename_with(~ gsub(version, "", .x)) |>
    clean_names()
  return(df)
}

process_currency <- function(df, lc_gc) {
  df <- df |>
    mutate(currency = lc_gc) |>
    relocate(c("currency"))
  return(df)
}

outlet <- function(filename) {
  # POC
  col_poc <- c("division", "bu", "new_outlet", "new_outlet_name")

  df <- read_excel(filename, range = cell_cols("A:F")) |>
    clean_names() |>
    select(!c("outlet_name")) |>
    select("outlet", all_of(col_poc))
  return(df)
}

join_with_outlet <- function(df, filename) {
  df <- df |>
    left_join(outlet(filename), by = "outlet")
  return(df)
}

profit_and_loss <- function(df) {
  df <- df |>
    select(!c("period_0", "ytd_0":"ytd_12")) |>
    filter(str_detect(.data[["financial_statement_item"]], "^3|^CO"))
  return(df)
}

balance_sheet <- function(df) {
  df <- df |>
    select(!c("period_0":"period_12")) |>
    filter(str_detect(.data[["financial_statement_item"]], "^1|^2"))
  return(df)
}

main <- function() {
  # Variables
  version <- "\r\nACT"

  # Filenames
  input_lc <- here(
    path, "data", "RACE",
    "Analysis FS Item Hierarchy for CU 698_LC.xlsx"
  )
  input_gc <- here(
    path, "data", "RACE",
    "Analysis FS Item Hierarchy for CU 698_GC.xlsx"
  )
  meta_file <- here(path, "meta", "New outlet.xlsx")
  output_pnl <- here(path, "output", "RACE Profit and Loss.csv")
  output_bs <- here(path, "output", "RACE Balance sheet.csv")


  # Combine data
  lc <- read_excel_file(input_lc, version) |> process_currency("LC")
  gc <- read_excel_file(input_gc, version) |> process_currency("GC")
  race <- bind_rows(lc, gc) |> join_with_outlet(meta_file)

  # Split data
  race_pnl <- profit_and_loss(race)
  race_bs <- balance_sheet(race)

  # Output data
  write_csv(race_pnl, output_pnl, na = "0")
  write_csv(race_bs, output_bs, na = "0")
}

main()
